R/quosure-constructor.R

```diff
@@ -0,0 +1,103 @@
+#' Initialize `Quosure` object
+#'
+#' Initialize `Quosure` object with `code` and `env`. In order to have `Quosure` reproducible
+#' one needs to initialize with `env` which can be reproduced by the `code`. Alternatively, one
+#' can create an empty `Quosure` and evaluate the expressions in this object using `eval_code`.
+#' @name new_quosure
+#'
+#' @param code (`character(1)` or `language`) code to evaluate. Accepts and stores comments also.
+#' @param env (`environment`) Environment being a result of the `code` evaluation.
+#'
+#' @examples
+#' new_quosure(env = list2env(list(a = 1)), code = "a <- 1")
+#' new_quosure(env = list2env(list(a = 1)), code = quote(a <- 1))
+#' new_quosure(env = list2env(list(a = 1)), code = parse(text = "a <- 1"))
+#'
+#' @export
+setGeneric("new_quosure", function(env = new.env(parent = parent.env(.GlobalEnv)), code = character(0)) {
+  standardGeneric("new_quosure")
+})
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure",
+  signature = c(env = "environment", code = "character"),
+  function(env, code) {
+    checkmate::assert_character(code, any.missing = FALSE)
+    new_env <- .copy_env(env)
+    lockEnvironment(new_env)
+
+    if (is.null(names(code)) && length(code)) {
+      names(code) <- make.unique(rep("initial code", length(code)))
+    }
+
+    code <- .keep_code_name_unique(code)
+    id <- sample.int(.Machine$integer.max, size = length(code))
+
+    methods::new("Quosure", env = new_env, code = code, id = id)
+  }
+)
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure",
+  signature = c(env = "environment", code = "expression"),
+  function(env, code) {
+    code_char <- as.character(code)
+    new_quosure(env = env, code = code_char)
+  }
+)
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure",
+  signature = c(env = "environment", code = "language"),
+  function(env, code) {
+    code_expr <- as.expression(code)
+    new_quosure(env = env, code = code_expr)
+  }
+)
+
+#' @rdname new_quosure
+#' @export
+setMethod(
+  "new_quosure",
+  signature = c(env = "list"),
+  function(env, code) {
+    if (!missing(code)) {
+      warning(
+        "`code` argument is ignored when creating Quosure from the list.",
+        "\nPlease pass the code througn attr(<list>, 'code')"
+      )
```

Comment:
why's that? `list` and `environment` are very close structures and the approach for passing on the code information differs.

Reply:
Marking as resolved:
discussed yesterday. We will make `data` as a new class and this data object will have wrapper get_code 
which will be used in the `new_quosure(list2env(data), get_code(data))`

https://github.com/insightsengineering/teal/issues/731#issue-1374679782